### PR TITLE
ReduceLp operation implementation

### PR DESCRIPTION
**Pull request by achetver**
Description: Current ReduceL2 operation was refactored to new ReduceLp operation

JIRA: 32991

Code:
* [x]  Comments
* [x]  Code style (PEP8)
* [x]  Transformation generates reshape-able IR
* [ ]  Transformation preserves original framework node names

Validation:
* [x]  Unit tests: Updated
* [x]  Framework operation tests: 35816
* [ ]  Transformation tests: N/A - no new transformation added
* [x]  e2e model test with an update of ./tests/e2e_oss/utils/reshape_utils.py: N/A - no new adeede models
* [x]  Model Optimizer IR Reader check: Done

Documentation:
* [x]  Supported frameworks operations list: Updated
* [x]  Supported **public** models list: No new supported models
* [x]  New operations specification: Done, PR: https://github.com/openvinotoolkit/openvino/pull/1205
* [x]  Guide on how to convert the **public** model: N/A - nothing to update
* [x]  User guide update: N/A - nothing to update

### CODE REVIEW

File: model-optimizer/extensions/ops/ReduceOps.py [hunk 1]
```diff
@@ -24,7 +24,7 @@
 reduce_map = {
     'ReduceSum': np.sum,
     'ReduceProd': np.prod,
-    'ReduceL2': lambda x, axis, keepdims: np.sqrt(np.sum(a=np.square(x), axis=axis, keepdims=keepdims)),
+    'ReduceLp': lambda x, axis, keepdims, p: np.power(np.sum(a=np.abs(np.power(x, p)), axis=axis, keepdims=keepdims), 1/p),
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/ops/ReduceOps.py:27]: pep8.

File: model-optimizer/extensions/front/reduce_axis_normalizer.py [hunk 2]
```diff
@@ -69,3 +70,10 @@ def replace_sub_graph(self, graph: Graph, match: [dict, SubgraphMatch]):
                 node.add_input_port(1, skip_if_exist=True)
                 axes.out_port(0).connect(node.in_port(1))
                 node.in_port(0).get_connection().get_source().connect(end_of_range.in_port(0))
+
+            # We need to add normalization order value for ReduceLp operation as 3rd input
+            if node.has('p'):
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/front/reduce_axis_normalizer.py:75]: `has` -> `has_valid`

Comment (lazarevevgeny) [inline model-optimizer/extensions/front/reduce_axis_normalizer.py:75]: This change is not related to the axis normalization. So I would recommend to move it to a separate transformation. Moreover, it is applicable for ReduceLp only while this one applies for all Reduce ops.

File: docs/MO_DG/prepare_model/Supported_Frameworks_Layers.md [hunk 3]
```diff
@@ -356,6 +356,8 @@ Standard ONNX\* operators:
 | ROIAlign | No |
 | Range | No |
 | Reciprocal | No |
+| ReduceL1 | No |
```
Comment (lazarevevgeny) [inline docs/MO_DG/prepare_model/Supported_Frameworks_Layers.md:359]: How about other frameworks? Do they have ReduceLx operations?

File: model-optimizer/extensions/front/ReduceL2Decomposition.py [hunk 4]
```diff
@@ -1,48 +0,0 @@
-"""
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/front/ReduceL2Decomposition.py:1]: Need to implement nGraph transformation performing the same decomposition.

File: model-optimizer/extensions/ops/ReduceOps.py [hunk 5]
```diff
@@ -138,11 +140,16 @@ class ReduceMean(ReduceOp):
     enabled = True
 
 
-class ReduceL2(ReduceOp):
-    op = 'ReduceL2'
-    op_type = None
+class ReduceLp(ReduceOp):
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/ops/ReduceOps.py:143]: Please, create a new operation AttributedReduceLp (where `p` is an attribute) and implement transformation which converts it to ReduceLp (with `p` as input).

File: model-optimizer/extensions/ops/ReduceOps.py (hunk 1 above)
Comment (achetver) [reply model-optimizer/extensions/ops/ReduceOps.py]: Done

File: model-optimizer/extensions/front/reduce_axis_normalizer.py (hunk 2 above)
Comment (achetver) [reply model-optimizer/extensions/front/reduce_axis_normalizer.py]: Separate transformation added.

File: docs/MO_DG/prepare_model/Supported_Frameworks_Layers.md (hunk 3 above)
Comment (achetver) [reply docs/MO_DG/prepare_model/Supported_Frameworks_Layers.md:368]: TensorFlow have `reduce_euclidean_norm` operation same as ReduceL2 operation from ONNX , extractor for `reduce_euclidean_norm` added.

File: model-optimizer/extensions/ops/ReduceOps.py [hunk 6]
```diff
@@ -138,11 +145,30 @@ class ReduceMean(ReduceOp):
...
-class ReduceL2(ReduceOp):
-    op = 'ReduceL2'
-    op_type = None
+class AttributedReduceLp(ReduceOp):
+    op = 'AttributedReduceLp'
+    op_type = 'AttributedReduceLp'
+    enabled = True
+
+    def __init__(self, graph: Graph, attrs: dict):
+        op_attrs = {'version': 'opset4',
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/ops/ReduceOps.py:154]: Let's add an assert statements to the constructor to check that both `axis` and `p` attributes are not `None`.

File: model-optimizer/extensions/front/AttributedReduceLpToReduceLp.py [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+            assert keep_dims is not None, \
+                'AttributedReduceLp should have `keep_dims` parameter set, but it`s not for node {}'.format(name)
+
+            reduce_node = create_op_with_const_inputs(graph, ReduceLp, {2: p}, {'name': name + '/ReduceLp',
+                                                                                'keep_dims': keep_dims})
+
+            node.in_port(0).get_connection().set_destination(reduce_node.in_port(0))
+            node.in_port(1).get_connection().set_destination(reduce_node.in_port(1))
+
+            for idx, port in node.out_ports().items():
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/front/AttributedReduceLpToReduceLp.py:49]: Lines 49-50 can be replaced with:

`node.out_port(0).get_connection().set_source(reduce_node.out_port(0))`

because the ReduceLp has just one output port.

File: model-optimizer/extensions/front/tf/reduce_ext.py [hunk 8]
```diff
@@ -67,3 +67,13 @@ class SumFrontExtractor(FrontExtractorOp):
     def extract(cls, node: Node):
         ReduceSum.update_node_stat(node, {'keep_dims': node.pb.attr["keep_dims"].b})
         return cls.enabled
+
+
+class EuclideanNormFrontExtractor(FrontExtractorOp):
+    op = 'EuclideanNorm'
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/front/tf/reduce_ext.py:73]: The TF EuclideanNorm operation has `axis` as input (not attribute). So we need another operation where `axis` is input.

File: docs/MO_DG/prepare_model/Supported_Frameworks_Layers.md [hunk 9]
```diff
@@ -365,6 +365,8 @@ Standard ONNX\* operators:
...
+| ReduceL2 | No |
 | ReduceMax | No |
 | ReduceMean | No |
```
Comment (lazarevevgeny) [inline docs/MO_DG/prepare_model/Supported_Frameworks_Layers.md:371]: Add `EuclideanNorm` to the list of supported TF operations.

File: model-optimizer/extensions/front/tf/reduce_ext.py (hunk 8 above)
Comment (achetver) [reply model-optimizer/extensions/front/tf/reduce_ext.py:73]: But in TF [specification](https://www.tensorflow.org/api_docs/python/tf/math/reduce_euclidean_norm) tf.math.reduce_euclidean_norm operation has `axis` as attribute, not as input

File: model-optimizer/extensions/ops/ReduceOps.py [hunk 10]
```diff
@@ -138,11 +145,30 @@ class ReduceMean(ReduceOp):
...
+                    'infer': None,
```
[the remark this answers is not in the document]
Comment (achetver) [reply model-optimizer/extensions/ops/ReduceOps.py:155]: In this case we break `ReduceAxisNormalizer` pass, when we connect constant node with `axis` as second input, and this pass common for all reduce nodes.

File: model-optimizer/extensions/front/AttributedReduceLpToReduceLp.py (hunk 7 above)
Comment (achetver) [reply model-optimizer/extensions/front/AttributedReduceLpToReduceLp.py]: Done.

File: model-optimizer/extensions/ops/ReduceOps.py (hunk 6 above)
Comment (achetver) [reply model-optimizer/extensions/ops/ReduceOps.py:157]: Done.

File: docs/MO_DG/prepare_model/Supported_Frameworks_Layers.md (hunk 9 above)
Comment (achetver) [reply docs/MO_DG/prepare_model/Supported_Frameworks_Layers.md:372]: Done.

File: model-optimizer/extensions/front/tf/reduce_ext.py (hunk 8 above)
Comment (lazarevevgeny) [reply model-optimizer/extensions/front/tf/reduce_ext.py:73]: From the specification that you mention it is not clear whether it is input or attribute. But if you take a look at the TF [source code](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/ops/math_ops.cc#L999) you will find the operation declaration:
```
REGISTER_OP("EuclideanNorm")
    .Input("input: T")
    .Input("reduction_indices: Tidx")
    .Output("output: T")
    .Attr("keep_dims: bool = false")
    .Attr("T: numbertype")
    .Attr("Tidx: {int32, int64} = DT_INT32")
    .SetShapeFn(shape_inference::ReductionShape);
```
and here you can see that `reduction_indices` is an input.

File: model-optimizer/extensions/ops/ReduceOps.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (lazarevevgeny) [reply model-optimizer/extensions/ops/ReduceOps.py:158]: The logic of converting the `AttributedReduceLp` to `ReduceLp` must be changed:

- Remove `AttributedReduceLp` from the `reduce_map`.
- Run the `AttributedReduceLpToReduceLp` before `ReduceAxisNormalizer`.

File: model-optimizer/extensions/front/AttributedReduceLpToReduceLp.py [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+
+            p = node.soft_get('p', None)
+            assert p is not None, \
+                'AttributedReduceLp should have `p` parameter set, but it`s not for node {}'.format(name)
+
+            keep_dims = node.soft_get('keep_dims', None)
+            assert keep_dims is not None, \
+                'AttributedReduceLp should have `keep_dims` parameter set, but it`s not for node {}'.format(name)
+
+            reduce_node = create_op_with_const_inputs(graph, ReduceLp, {2: p}, {'name': name + '/ReduceLp',
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/front/AttributedReduceLpToReduceLp.py:43]: The transformation must keep the original node name.

File: model-optimizer/extensions/front/AttributedReduceLpToReduceLp.py [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+"""
+
+from extensions.front.reduce_axis_normalizer import ReduceAxisNormalizer
+from extensions.ops.ReduceOps import ReduceLp
+from mo.front.common.replacement import FrontReplacementSubgraph
+from mo.front.tf.graph_utils import create_op_with_const_inputs
+from mo.graph.graph import Graph
+
+
+class AttributedReduceLpToReduceLp(FrontReplacementSubgraph):
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/front/AttributedReduceLpToReduceLp.py:24]: I don't see the unit test for the transformation

**Comment by lazarevevgeny** [comment]
The work on this task will be completed in https://github.com/openvinotoolkit/openvino/pull/1799
